### PR TITLE
Convert keys to binary in the Redis cache store

### PR DESCRIPTION
**Pull request by georgeclaghorn**
Fix encoding errors when using the pure-Ruby Redis driver instead of Hiredis. Dodge incompatibilities between UTF-8 and arbitrary value encodings, which rear their heads when the Redis driver tries to build a single command string from incompatibly-encoded keys and values.

/cc @jeremy

### CODE REVIEW

**Comment by kaspth** [comment]
Curious, what the benefit of the pure Ruby Redis client is over hiredis? Is Basecamp testing it out to replace hiredis?

**Comment by georgeclaghorn** [comment]
We don’t use Hiredis currently. I don’t know of any particular reason why; we probably just never set it up.

File: activesupport/lib/active_support/cache/redis_cache_store.rb [hunk 1]
```diff
@@ -363,7 +363,15 @@ def write_multi_entries(entries, expires_in: nil, **options)
...
-          truncate_key super
+          truncate_key encode_key(super)
+        end
+
+        def encode_key(key)
+          if key.encoding == Encoding::BINARY
+            key
+          else
+            key.dup.force_encoding(Encoding::BINARY)
+          end
```
Comment (jeremy) [inline activesupport/lib/active_support/cache/redis_cache_store.rb]: Perhaps we could get away with just `key.b` to dup & strip encoding? (Binary keys should be uncommon, given default source encoding, so the no-op may not save us much if anything.)

**Review by jeremy — APPROVED** [review]

File: activesupport/lib/active_support/cache/redis_cache_store.rb (hunk 1 above)
Comment (georgeclaghorn) [reply activesupport/lib/active_support/cache/redis_cache_store.rb]: Didn’t know about `String#b`. Switched.

**Review by rafaelfranca — APPROVED** [review]

File: .travis.yml [hunk 2]
```diff
@@ -70,6 +70,10 @@ matrix:
   include:
     - rvm: 2.5.0
       env: "GEM=av:ujs"
+    - rvm: 2.5.0
+      env: "GEM=as:cache:redis REDIS=ruby"
+      services:
+        - redis-server
```
[the remark this answers is not in the document]
Comment (matthewd) [reply .travis.yml]: Each test job takes forever to start.. is there any practical way we can fold this into the existing run?

[the remark this answers is not in the document]
Comment (georgeclaghorn) [reply .travis.yml]: I don’t know of one, short of dynamically generating test classes for each driver. Does that seem unreasonable?

[the remark this answers is not in the document]
Comment (matthewd) [reply .travis.yml]: Not necessarily: I think part of the AJ suite does something along those lines.

And we have a smaller scale version in the file watcher -- the tests are defined in a module, then the two driver-specific test classes just include said module and override what they must.

[the remark this answers is not in the document]
Comment (georgeclaghorn) [reply .travis.yml]: Active Job’s setup seems like what we want. I’ll take a stab at that.

**Comment by georgeclaghorn** [comment]
@matthewd, I switched to something resembling Active Job’s setup, which resulted in a big simplification. I’m pretty happy with this.

**Pull request merged**